### PR TITLE
[Documentation] New API example Notebook 1

### PR DESCRIPTION
**Pull request by zhjwy9343**
*Issue #, if available:*

*Description of changes:*
This PR create a new Jupyter notebook to demonstrate the general APIs for GraphStorm pipeline in a NC task. It also fixed a few issues with existing code that not work for the notebook.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 1]
```diff
@@ -0,0 +1,407 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Notebook 1: Use GraphStorm APIs for a Node Classification Pipeline\n",
+    "\n",
+    "This notebook demonstrates how to use GraphStorm's APIs to create graph machine learning pipelines. By playing with this notebook, users will be able to get familiar with GraphStorm pipeline APIs, hence further using them on their own tasks and models.\n",
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:9]:
```suggestion
    "This notebook demonstrates how to use GraphStorm's APIs to create graph machine learning pipelines. By playing with this notebook, users will be able to get familiar with GraphStorm APIs, hence further using them on their own tasks and models.\n",
```
pipeline APIs or APIs?

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    "gs.initialize()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1. Setup GraphStorm Dataset and DataLoaders\n",
+    "\n",
+    "Similar as Pytorch model training pipeline, we create dataset by constructing `gs.dataset.GSgnnData` class. In most case, users only need to provide the location of the graph description JSON file, which is created in GraphStorm's gconstruct operation."
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:63]:
```suggestion
    "Similar as Pytorch model training pipeline, we create a dataset by constructing `gs.dataset.GSgnnData` class. In most cases, users only need to provide the location of the graph description JSON file, which is created in GraphStorm's gconstruct operation."
```

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+   "source": [
+    "# create a GraphStorm Dataset for the ACM graph data generated in the Notebook 0\n",
+    "acm_data = gs.dataloading.GSgnnData(part_config='./acm_gs_1p/acm.json')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Then, we create different `DataLoader` for training, validation, and testing. As shown below, we allow users to specify different `DataLoader` settings, e.g., `fanout`, `batch_size`, except for a few model-related properties, such as `node_feats` and `label_field`.\n",
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:82]:
```suggestion
    "Then, we create different `DataLoader`s for training, validation, and testing. As shown below, we allow users to specify different `DataLoader` settings, e.g., `fanout`, `batch_size`, except for a few model-related properties, such as `node_feats` and `label_field`.\n",
```

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# define dataloaders for training and validation\n",
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:95]:
```suggestion
    "# define dataloaders for training, validation amd testing\n",
```

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# setup a Classification evaluator for the trainer\n",
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:170]:
```suggestion
    "# setup a classification evaluator for the trainer\n",
```

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+    "            save_model_path='a_save_path/')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### (Optional) 5. Visualize Model Performance History\n",
+    "\n",
+    "Besides the log, we can examine the model performance on the validation, and testing by visualizing evalutors' history properties."
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:220]:
```suggestion
    "Besides the log, we can examine the model performance on the validation, and testing by visualizing evalutors' history property."
```

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
...hh8x5//HF0Ot1Nt169emldmhDCTslmqEIIu9CrVy++++67YsdcXV01qkYIYe+kBUgIYRdcXV3x9/cvdqtatSqgdk99/fXX9O7dG3d3d+rWrcvPP/9c7PGHDh2iW7duuLu7U716dZ566imysrKKnTNnzhwaNWqEq6srAQEBjBs3rtjP09LSePDBB/Hw8KB+/fqsXLnSui9aCGE1EoCEEBXCG2+8wcMPP8yBAwcYNmwYQ4YM4dixYwBkZ2fTs2dPqlatyu7du1myZAlr164tFnC+/vprxo4dy1NPPcWhQ4dYuXIloaGhxZ7j7bffZtCgQRw8eJA+ffowbNgwLl26VK6vUwhhIVpvRy+EEP9k5MiRisFgUDw9PYvd3nvvPUVRFAVQnn766WKPiYyMVJ555hlFURTlm2++UapWrapkZWWZf75q1SpFr9crSUlJiqIoSmBgoPLaa6/dsgZAef31183fZ2VlKYDy+++/W+x1CiHKj4wBEkLYha5du/L1118XO1atWjXz/fbt2xf7Wfv27dm/fz8Ax44do1mzZnh6epp/3rFjR0wmEzExMeh0Os6fP0/37t1vW0PTpk3N9z09PfHy8iIlJeVOX5IQQkMSgIQQdsHT0/OmLilLcXd3L9V5zs7Oxb7X6XSYTCZrlCSEsDIZAySEqBB27Nhx0/fh4eEAhIeHc+DAAbKzs80/37p1K3q9nrCwMCpXrkxISAjR0dHlWrMQQjvSAiSEsAu5ubkkJSUVO+bk5ISPjw8AS5YsoXXr1nTq1IkffviBXbt2MXv2bACGDRvG5MmTGTlyJG+99RapqamMHz+e4cOH4+fnB8Bbb73F008/ja+vL7179yYzM5OtW7cyfvz48n2hQohyIQFICGEXVq9eTUBAQLFjYWFhHD9+HFBnaC1atIhnn32WgIAAfvzxRyIiIgDw8PDgjz/+YMKECbRp0wYPDw8efvhhpk6dar7WyJEjycnJ4bPPPuPFF1/Ex8eHRx55pPxeoBCiXOkURVG0LkIIIe6GTqdj2bJlDBgwQOtShBB2QsYACSGEEMLhSAASQgghhMORMUBCCLsnPflCiLKSFiAhhBBCOBwJQEIIIYRwOBKAhBBCCOFwJAAJIYQQwuFIABJCCCGEw5EAJIQQQgiHIwFICCGEEA5HApAQQgghHM7/A2PSCzifVWGXAAAAAElFTkSuQmCC",
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:242]: Do we need to include the generated image in the PR?

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+    "ax.legend(loc='best')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 6. Inference with the Trained Model\n",
+    "\n",
+    "GraphStorm automatically save the best performaned model according to the value specified in the `save_model_path` argument. We can first find out what is the best model and its path. And then restore it by using model's `restore_model()` method."
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:274]:
```suggestion
    "GraphStorm automatically save the best performed model according to the value specified in the `save_model_path` argument. We can first find out what is the best model and its path. And then restore it by using model's `restore_model()` method."
```

File: docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+    "                                                      fanout=[100, 100],\n",
+    "                                                      batch_size=256,\n",
+    "                                                      train_task=False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "GraphStorm provides a set of inferrers that can perform highly efficient inference for very large graphs. Similar to trainers, users can create inferrers by given models as input argument, and then call its `infer()` method with a few inference related parameters, such as folder paths for saving inference results."
```
Comment (classicsong) [inline docs/source/notebooks/Notebook_1_NC_Pipeline.ipynb:335]:
```suggestion
    "GraphStorm provides a set of inferrers that can perform highly efficient inference for very large graphs. Similar to trainers, users can create inferrers by giving a model as its input argument, and then call its `infer()` method with a few inference related parameters, such as folder paths for saving inference results."
```

File: docs/source/notebooks/demo_models.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
+"""
```
Comment (classicsong) [inline docs/source/notebooks/demo_models.py:1]: The license should look like: https://github.com/awslabs/graphstorm/blob/main/tests/end2end-tests/data_process/test_custom_mask_data.py#L1-L16

File: docs/source/notebooks/demo_models.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        # classification loss function
+        self.set_loss_func(ClassifyLossFunc(multilabel=multilabel))
+
+        # initialize model's optimizer
+        self.init_optimizer(lr=0.001,
+                            sparse_optimizer_lr=0.01,
+                            weight_decay=0)
+
+
+class RgcnLPModel(GSgnnLinkPredictionModel):
```
Comment (classicsong) [inline docs/source/notebooks/demo_models.py:100]: Is RgcnLPModel used in this PR? If not, let's remove it.
We can add it in the next PR.

File: python/graphstorm/eval/evaluator.py [hunk 12]
```diff
@@ -797,6 +797,7 @@ def evaluate(self, val_rankings, test_rankings, total_iters):
             else:
                 for metric in self.metric_list:
                     val_score = {metric: "N/A"} # Dummy
+        self._history.append((val_score, test_score))
```
Comment (classicsong) [inline python/graphstorm/eval/evaluator.py:800]: Is this a bug? Maybe we should open a standalone PR for fixing this.

File: tests/end2end-tests/graphstorm-lp/test.sh [hunk 13]
```diff
@@ -38,7 +38,7 @@ python3 -m graphstorm.run.gs_link_prediction --workspace $GS_HOME/training_scrip
 error_and_exit $?
 
 echo "**************dataset: Movielens, RGCN layer 1, node feat: fixed HF BERT & construct, BERT nodes: movie, inference: mini-batch, negative_sampler: joint, exclude_training_targets: false"
-python3 -m graphstorm.run.gs_link_prediction --workspace $GS_HOME/training_scripts/gsgnn_lp --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_lp_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_lp.yaml --num-epochs 1 --eval-frequency 300 --node-feat-name movie:title --construct-feat-ntype user
+python3 -m graphstorm.run.gs_link_prediction --workspace $GS_HOME/training_scripts/gsgnn_lp --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_lp_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_lp.yaml --num-epochs 1 --eval-frequency 300 --node-feat-name movie:title --construct-feat-ntype user --use-mini-batch-infer false
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-lp/test.sh:41]: Why do we want to change the test?

File: python/graphstorm/trainer/lp_trainer.py [hunk 14]
```diff
@@ -128,6 +129,14 @@ def fit(self, train_loader, num_epochs,
         if not use_mini_batch_infer:
             assert isinstance(self._model, GSgnnModel), \
                     "Only GSgnnModel supports full-graph inference."
+
+        # assert not use GNNEncoderWithReconstructedEmbed in use_mini_batch_infer=True
+        if self._model.gnn_encoder is not None:
```
Comment (classicsong) [inline python/graphstorm/trainer/lp_trainer.py:134]: We should have a separate PR for fixing this.

File: python/graphstorm/trainer/lp_trainer.py [hunk 15]
```diff
@@ -128,6 +129,14 @@ def fit(self, train_loader, num_epochs,
         if not use_mini_batch_infer:
             assert isinstance(self._model, GSgnnModel), \
                     "Only GSgnnModel supports full-graph inference."
+
+        # assert not use GNNEncoderWithReconstructedEmbed in use_mini_batch_infer=True
```
Comment (classicsong) [inline python/graphstorm/trainer/lp_trainer.py:133]:
```suggestion
        # assert not use GNNEncoderWithReconstructedEmbed when use_mini_batch_infer=True
```

File: python/graphstorm/trainer/ep_trainer.py [hunk 16]
```diff
@@ -47,7 +47,7 @@ class GSgnnEdgePredictionTrainer(GSgnnTrainer):
     model : GSgnnEdgeModel
         The GNN model for edge prediction.
     topk_model_to_save : int
-        The top K model to save.
+        The top K model to save. Default: 1
```
Comment (classicsong) [inline python/graphstorm/trainer/ep_trainer.py:50]: This is an API change. I think we should have a separate PR for this.

File: python/graphstorm/gsf.py [hunk 17]
```diff
@@ -88,6 +88,14 @@
 from .dataloading import (GSgnnLinkPredictionTestDataLoader,
                           GSgnnLinkPredictionJointTestDataLoader,
                           GSgnnLinkPredictionPredefinedTestDataLoader)
+from .trainer import (GSgnnLinkPredictionTrainer,
```
Comment (classicsong) [inline python/graphstorm/gsf.py:91]: Why the Trainers and Inferrers are imported by gsf.py?
Are we using these classes?

**Comment by classicsong** [comment]
Can we close this PR?

**Comment by zhjwy9343** [comment]
This PR had been divided into 4 separated PRs. Therefore, close it for merging.

**Pull request closed without merging**